test/routes/api/tasks/tasks.test.ts

```diff
@@ -457,182 +471,277 @@ describe('Tasks api (logged user only)', () => {
+
+        assert.deepStrictEqual(arg.err.message, 'Transaction failed.')
       })
+    })
+    
+    describe('Retrieval', () => {
+      it('should retrieve the uploaded image based on task id and filename', async (t) => {
+        app = await build(t)
+
+        const taskFilename = encodeURIComponent(`${taskId}_${filename}`)
+        const res = await app.injectWithLogin('basic', {
+          method: 'GET',
+          url: `/api/tasks/${taskFilename}/image`
+        })
 
-      const { mock: mockLogError } = t.mock.method(app.log, 'error')
+        assert.strictEqual(res.statusCode, 200)
+        assert.strictEqual(res.headers['content-type'], 'image/png')
 
-      const form = new FormData()
-      form.append('file', fs.createReadStream(testImagePath))
-      const res = await app.injectWithLogin('basic', {
-        method: 'POST',
-        url: `/api/tasks/${taskId}/upload`,
-        payload: form,
-        headers: form.getHeaders()
+        const originalFile = fs.readFileSync(testImagePath)
+
+        assert.deepStrictEqual(originalFile, res.rawPayload)
       })
 
-      assert.strictEqual(res.statusCode, 500)
-      assert.strictEqual(mockLogError.callCount(), 1)
+      it('should return 404 error for non-existant filename', async (t) => {
+        app = await build(t)
 
-      const arg = mockLogError.calls[0].arguments[0] as unknown as {
-        err: Error;
-      }
+        const res = await app.injectWithLogin('basic', {
+          method: 'GET',
+          url: '/api/tasks/non-existant/image'
+        })
 
-      assert.deepStrictEqual(arg.err.message, 'Transaction failed.')
+        assert.strictEqual(res.statusCode, 404)
+        const { message } = JSON.parse(res.payload)
+        assert.strictEqual(message, 'No task has filename "non-existant"')
+      })
     })
 
-    it('should retrieve the uploaded image based on task id and filename', async (t) => {
-      app = await build(t)
+    describe('Deletion', () => {
+      before(async () => { 
+        app = await build()
 
-      const taskFilename = encodeURIComponent(`${taskId}_${filename}`)
-      const res = await app.injectWithLogin('basic', {
-        method: 'GET',
-        url: `/api/tasks/${taskFilename}/image`
+        await app.knex<Task>('tasks').where({ id: taskId }).update({ filename: null})
+        
+        const files = fs.readdirSync(uploadDirTask)
+        files.forEach((file) => {
+          const filePath = path.join(uploadDirTask, file)
+          fs.rmSync(filePath, { recursive: true })
+        })       
+
+        await app.close()
       })
 
-      assert.strictEqual(res.statusCode, 200)
-      assert.strictEqual(res.headers['content-type'], 'image/png')
+      beforeEach(async () => {
+        app = await build()
+        await uploadImageForTask(app, taskId, testImagePath, uploadDirTask)
+        await app.close()
+      })
 
-      const originalFile = fs.readFileSync(testImagePath)
+      after(async () => {
+        app = await build()
```

Comment:
Why do you build the app here? 
Here you are building to app but only closing one, maybe this is the reason of you bug.
If you don't pass `t` (TestContext) to build, it will not close automatically, this is why we need to close it in `after` hook.

Reply:
i have a really weird behavior with this bug, i just deleted the build and the close in the after hook, make some changes for other comments, all works fine, the bug disappeared, then i added a another tests for 100% coverage (the 404 file note found for the unlink) and then the bug appeared again.

Reply:
When yo uremove test `should return 404 for non-existant file in upload dir`, it does work?

Reply:
Yes exactly !